Overhaul the plugin cache file with a new msgpack+brotli format

devyn opened the pull request:
# Description

- Plugin signatures are now saved to `plugin.msgpackz`, which is brotli-compressed MessagePack.
- The file is updated incrementally, rather than writing all plugin commands in the engine every time.
- The file always contains the result of the `Signature` call to the plugin, even if commands were removed.
- Invalid data for a particular plugin just causes an error to be reported, but the rest of the plugins can still be parsed

# User-Facing Changes

- The plugin file has a different filename, and it's not a nushell script.
- The default `plugin.nu` file will be automatically migrated the first time, but not other plugin config files.
- We don't currently provide any utilities that could help edit this file, beyond `plugin add` and `plugin rm`
  - `from msgpackz`, `to msgpackz` could also help
- New commands: `plugin add`, `plugin rm`

# Tests + Formatting

Tests added for the format and for the invalid handling.

- :green_circle: `toolkit fmt`
- :green_circle: `toolkit clippy`
- :green_circle: `toolkit test`
- :green_circle: `toolkit test stdlib`

# After Submitting

- [ ] Check for documentation changes
- [ ] Definitely needs release notes

fdncred commented on crates/nu-cli/src/config_files.rs:
```diff
@@ -38,38 +41,95 @@ pub fn read_plugin_file(
+                        msg: "plugin path defined here".into(),
+                        span,
+                        help: Some(
+                            "you might try deleting the file and registering all of your \
```
Not sure about this `\` since nushell doesn't support line continuation characters like this.

devyn replied:
this is a Rust syntax feature for wrapping strings

fdncred replied:
ah, right. thx.

fdncred commented on crates/nu-protocol/Cargo.toml:
```diff
@@ -25,19 +26,25 @@ indexmap = { workspace = true }
 strum = "0.26"
 strum_macros = "0.26"
 nu-test-support = { path = "../nu-test-support", version = "0.92.3" }
+pretty_assertions = "1.0"
```
I'm wondering if pretty_assertions is necessary? We try to keep dependencies to a minimum. I'm just not sure of the value here.

devyn replied:
It's just for the tests, we're using it in other places, and it's a very large thing to `assert_eq!` that I'd probably otherwise have to pull apart and compare individual pieces of to be useful

fdncred commented:
Thanks devyn. This looks like a good start to me. I'm wondering if a special `plugin migrate` would be helpful to users for a limited time, or like you say above, should there be an auto-migrate? Do you plan for the other commands to be in this PR or will they be separate? If separate, we'll need to land them close to each other so people can use it.

devyn commented:
> Thanks devyn. This looks like a good start to me. I'm wondering if a special `plugin migrate` would be helpful to users for a limited time, or like you say above, should there be an auto-migrate? Do you plan for the other commands to be in this PR or will they be separate? If separate, we'll need to land them close to each other so people can use it.

I think auto-migrate at startup will be possible, I just have to `eval_source()` the plugin.nu if it exists and plugin.msgpackz doesn't, and then group and write out the defs, then print a message to stderr.

I think I agree with adding at least `plugin add` and `plugin rm` to this PR, and maybe I'll do `plugin use` and deprecate `register` separately.

devyn commented:
- Added `plugin add` and `plugin rm`
- Moved plugin commands to a `nu-cmd-plugin` crate to separate the dependencies
- Added `Plugin` category for the plugin commands (including `register`)
- Changed buffer size for brotli to 64k

Now able to load a plugin file containing all of the polars stuff in 5.8ms

@fdncred it turns out brotli does add about 50-80% on top, I get more like 3ms without brotli. since it's also an extra bunch of code to include, and the files aren't that big anyway, maybe we shouldn't?

fdncred commented on crates/nu-cmd-plugin/src/commands/plugin/add.rs:
```diff
@@ -0,0 +1,148 @@
+                result: None,
+            },
+            Example {
+                example: "plugin add --plugin-config=polars.msgpackz nu_plugin_polars",
```
I'm not a super fan of this syntax `--plugin-config=polars.msgpackz`. Can we do the same thing but without the equal sign?

devyn commented:
Ready to go (probably).

The migration has been added, so please test that as well. I did a really basic message on stderr if the shell is running in interactive mode.

kubouch commented:
With `nu_plugin_polars` and a few other plugins (like gstat):

Before:
```
read plugin file /home/kubouch/.config/nushell/plugin.nu took 184.831µs
eval_source /home/kubouch/.config/nushell/plugin.nu took 64.304891ms
```

After:
```
read plugin file /home/kubouch/.config/nushell/plugin.msgpackz took 3.611742ms
load plugin file /home/kubouch/.config/nushell/plugin.msgpackz took 98.06µs
```

This is wild!

fdncred commented:
> @fdncred it turns out brotli does add about 50-80% on top, I get more like 3ms without brotli. since it's also an extra bunch of code to include, and the files aren't that big anyway, maybe we shouldn't?

I'm good with whatever. I just like msgpack + brotli but I'm fine without it as well. If we go without it, just make sure the file isn't named msgpackz.

devyn commented:
Let's keep brotli in then. Having a small file is nice, it's a binary format anyway, and there's a lot of repetition in it that's easily compressed

fdncred commented:
Just found something else that needs to be updated, toolkit.nu. I do this all the time.
```nushell
use toolkit.nu
toolkit register plugins
```

fdncred commented:
A nice-to-have would be to be able to use `plugin rm nu_plugin_msgpack` and it remove all the signatures associated with the plugin file. 

My first attempt was `plugin rm nu_plugin_msgpack` since it seemed analogous with `plugin add nu_plugin_msgpack`. Nope, that didn't work.

Then I saw this Example.
```nushell
  Remove the installed signatures for the `inc` plugin.
  > plugin rm inc
```

So, I tried `plugin rm 'from msgpack'` and `plugin rm 'to msgpack'`, but of course that didn't work. I finally saw `name <string>: The name of the plugin to remove (not the filename)` which clued me in that I needed to do `plugin rm msgpack`. 

It's a little confusing. I'm fine with leaving it but it would be nice to have the plugin file name too because it would go hand-in-hand with `plugin add nu_plugin_msgpack`.

fdncred commented:
Sorry to keep adding nice-to-haves, but here's another one `plugin add /path/to/plugins`. I have 22 plugins right now and it would be nice to just have it iterate a folder looking for nu_plugin_* and register them. Kind of goes to the reason why I use `toolkit register plugins`.

devyn commented:
@fdncred 

Since `plugin add` isn't a keyword, this should be possible now:

```nushell
glob /path/to/plugins/nu_plugin_* | each { |bin| plugin add $bin }
```

You could even generate a string of `plugin use` commands if you wanted to be able to copy+paste that too
